### PR TITLE
Add file/info command

### PR DESCRIPTION
**Pull request by 4e6**
### Pull Request Description

closes: #540 

PR adds `file/info` filemanager command.

ChangeLog:
- add: `info` filesystem command
- add: `FileAttributes` type to protocol
- add: `UTCDateTime` type to protocol

### Important Notes

External protocol extended:

```typescript
/**
 * A representation of the attributes of a file.
 *
 * @param creationTime creation time
 * @param lastAccessTime last access time
 * @param lastModifiedTime last modified time
 * @param kind type of [[FileSystemObject]], can be:
 * `DirectoryTruncated`, `File`, `Other`
 * @param byteSize size in bytes
 */
interface FileAttributes {
  creationTime: UTCDateTime;
  lastAccessTime: UTCDateTime;
  lastModifiedTime: UTCDateTime;
  kind: FileSystemObject;
  byteSize: number;
}

/**
 * Time in UTC time zone represented as ISO-8601 string
 */
type UTCDateTime = String;

```

### Checklist
Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Scala](https://github.com/luna/enso/blob/master/doc/scala-style-guide.md), [Java](https://github.com/luna/enso/blob/master/doc/java-style-guide.md), [Rust](https://github.com/luna/enso/blob/master/doc/rust-style-guide.md) or [Haskell](https://github.com/luna/enso/blob/master/doc/haskell-style-guide.md) style guides as appropriate.
- [x] All code has been tested where possible.

### CODE REVIEW

**Comment by 4e6** [comment]
I thought a number would be easier to parse. We can ask ide guys what format they prefer

**Comment by 4e6** [comment]
ChangeLog:
- update: convert time to `OffsetDateTime`
- update: JSON `UTCDateTime` representation to an ISO-8601 string
  ``` typescript
  /**
   * Time in UTC time zone represented as ISO-8601 string
   */
  type UTCDateTime = String;
  ```

File: engine/language-server/src/main/scala/org/enso/languageserver/filemanager/FileSystem.scala [hunk 1]
```diff
@@ -198,6 +199,25 @@ class FileSystem extends FileSystemApi[BlockingIO] {
...
+    */
+  override def info(
+    path: File
+  ): BlockingIO[FileSystemFailure, Attributes] =
+    if (path.exists) {
+      effectBlocking({
+        val attrs =
+          Files.readAttributes(path.toPath, classOf[BasicFileAttributes])
+        Attributes.fromBasicAttributes(path.toPath, attrs)
+      }).mapError(errorHandling)
```
Comment (lolczak) [inline engine/language-server/src/main/scala/org/enso/languageserver/filemanager/FileSystem.scala]: Do we need here both round and curly brackets?

File: engine/language-server/src/main/scala/org/enso/languageserver/filemanager/FileSystemApi.scala [hunk 2]
```diff
@@ -163,4 +185,53 @@ object FileSystemApi {
...
+      Attributes(
+        creationTime     = basic.creationTime(),
+        lastAccessTime   = basic.lastAccessTime(),
+        lastModifiedTime = basic.lastModifiedTime(),
+        kind             = Entry.fromBasicAttributes(path, basic),
+        byteSize         = basic.size()
+      )
+
+    private def utcTime(time: FileTime): OffsetDateTime =
+      OffsetDateTime.ofInstant(time.toInstant, ZoneId.of("UTC"))
```
Comment (lolczak) [inline engine/language-server/src/main/scala/org/enso/languageserver/filemanager/FileSystemApi.scala]: You can use `ZoneOffset.UTC` here, it's faster.

File: engine/language-server/src/main/scala/org/enso/languageserver/filemanager/UTCDateTime.scala [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+        Try(OffsetDateTime.parse(string, IsoDateTime)).toEither
+          .map(UTCDateTime(_))
+          .left
+          .map(DecodingFailure.fromThrowable(_, cursor.history))
+      }
+    }
+
+  implicit val encoder: Encoder[UTCDateTime] =
+    Encoder.instance[UTCDateTime](_.time.format(IsoDateTime).asJson)
+}
```
Comment (lolczak) [inline engine/language-server/src/main/scala/org/enso/languageserver/filemanager/UTCDateTime.scala]: There is no point in wrapping `OffsetDateTime` in `UTCDateTime` and write decoders and encoders for it. Circe has default encoder for the `OffsetDateTime` that produces ISO-8601 format.

File: engine/language-server/src/main/scala/org/enso/languageserver/requesthandler/file/InfoFileHandler.scala [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+import org.enso.languageserver.filemanager.{
+  FileManagerProtocol,
+  FileSystemFailureMapper
+}
+import org.enso.languageserver.filemanager.FileManagerApi.InfoFile
+import org.enso.languageserver.requesthandler.RequestTimeout
+
+import scala.concurrent.duration.FiniteDuration
+
+class InfoFileHandler(requestTimeout: FiniteDuration, fileManager: ActorRef)
```
Comment (lolczak) [inline engine/language-server/src/main/scala/org/enso/languageserver/requesthandler/file/InfoFileHandler.scala]: Scala docs are missing here

File: engine/language-server/src/main/scala/org/enso/languageserver/requesthandler/file/InfoFileHandler.scala [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+
+    case FileManagerProtocol.InfoFileResult(Right(result)) =>
+      replyTo ! ResponseResult(InfoFile, id, InfoFile.Result(result))
+      context.stop(self)
+  }
+}
+
+object InfoFileHandler {
+
+  def props(timeout: FiniteDuration, fileManager: ActorRef): Props =
```
Comment (lolczak) [inline engine/language-server/src/main/scala/org/enso/languageserver/requesthandler/file/InfoFileHandler.scala]: Scala docs are missing here

File: engine/language-server/src/main/scala/org/enso/languageserver/filemanager/FileSystemApi.scala [hunk 6]
```diff
@@ -163,4 +185,53 @@ object FileSystemApi {
...
+    ): Attributes =
+      Attributes(
+        creationTime     = utcTime(creationTime),
+        lastAccessTime   = utcTime(lastAccessTime),
+        lastModifiedTime = utcTime(lastModifiedTime),
+        kind             = kind,
+        byteSize         = byteSize
+      )
+
+    def fromBasicAttributes(
```
Comment (lolczak) [inline engine/language-server/src/main/scala/org/enso/languageserver/filemanager/FileSystemApi.scala]: Scala docs are missing here

File: engine/language-server/src/main/scala/org/enso/languageserver/filemanager/FileSystemApi.scala [hunk 7]
```diff
@@ -163,4 +185,53 @@ object FileSystemApi {
...
+    creationTime: OffsetDateTime,
+    lastAccessTime: OffsetDateTime,
+    lastModifiedTime: OffsetDateTime,
+    kind: Entry,
+    byteSize: Long
+  )
+
+  object Attributes {
+
+    def apply(
```
Comment (lolczak) [inline engine/language-server/src/main/scala/org/enso/languageserver/filemanager/FileSystemApi.scala]: Scala docs are missing here

File: engine/language-server/src/main/scala/org/enso/languageserver/filemanager/FileAttributes.scala [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+  creationTime: UTCDateTime,
+  lastAccessTime: UTCDateTime,
+  lastModifiedTime: UTCDateTime,
+  kind: FileSystemObject,
+  byteSize: Long
+)
+
+object FileAttributes {
+
+  def fromFileSystemAttributes(
```
Comment (lolczak) [inline engine/language-server/src/main/scala/org/enso/languageserver/filemanager/FileAttributes.scala]: Scala docs are missing here

File: engine/language-server/src/main/scala/org/enso/languageserver/filemanager/FileSystemApi.scala [hunk 9]
```diff
@@ -121,6 +131,18 @@ object FileSystemApi {
     def path: Path
   }
 
+  object Entry {
+
+    def fromBasicAttributes(path: Path, attrs: BasicFileAttributes): Entry =
```
Comment (lolczak) [inline engine/language-server/src/main/scala/org/enso/languageserver/filemanager/FileSystemApi.scala]: Scala docs are missing here

File: engine/language-server/src/main/scala/org/enso/languageserver/filemanager/Path.scala [hunk 10]
```diff
@@ -44,4 +44,6 @@ object Path {
   def getRelativeParent(root: File, base: Path, path: nio.file.Path): Path =
     getRelativePath(root, base, path.getParent())
 
+  def getRelativeParent(root: File, path: Path): Path =
```
Comment (lolczak) [inline engine/language-server/src/main/scala/org/enso/languageserver/filemanager/Path.scala]: Scala docs are missing here

File: engine/language-server/src/main/scala/org/enso/languageserver/filemanager/UTCDateTime.scala [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+  * A representation of UTC time.
+  *
+  * @param time time in UTC zone
+  */
+case class UTCDateTime(time: OffsetDateTime)
+
+object UTCDateTime {
+
+  private val IsoDateTime =
+    DateTimeFormatter.ISO_ZONED_DATE_TIME
```
Comment (lolczak) [inline engine/language-server/src/main/scala/org/enso/languageserver/filemanager/UTCDateTime.scala]: It's better to use `DateTimeFormatter.ISO_OFFSET_DATE_TIME` here.

**Comment by 4e6** [comment]
ChangeLog:

- remove: `UTCDateTime`
- add: docs

File: engine/language-server/src/main/scala/org/enso/languageserver/filemanager/FileSystem.scala (hunk 1 above)
Comment (lolczak) [reply engine/language-server/src/main/scala/org/enso/languageserver/filemanager/FileSystem.scala]: Please remove round brackets here.

File: engine/language-server/src/main/scala/org/enso/languageserver/filemanager/FileAttributes.scala (hunk 8 above)
Comment (lolczak) [reply engine/language-server/src/main/scala/org/enso/languageserver/filemanager/FileAttributes.scala]: Please add scala doc here

File: engine/language-server/src/main/scala/org/enso/languageserver/filemanager/FileSystemApi.scala (hunk 2 above)
Comment (lolczak) [reply engine/language-server/src/main/scala/org/enso/languageserver/filemanager/FileSystemApi.scala]: Please use ZoneOffset.UTC.

**Pull request merged**